### ISSUE TITLE
Epic 2: Minimal Learned Core

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#5]
**Goal**: Train 20M param Graph-to-Graph Transformer for AST reasoning

**Duration**: Month 3-4

**Success Criteria**:
- [ ] Model: 20M params total (320K root embeddings + 19.68M reasoning)
- [ ] Accuracy: 70%+ on multi-hop reasoning (200 test questions)
- [ ] Improvement: +15% accuracy over deterministic-only baseline
- [ ] Inference speed: <50ms per query

**Components to Build**:
1. Design Graph-to-Graph Transformer architecture (input AST → output AST)
2. Implement root embeddings ONLY (5K roots × 64d = 320K params)
3. **KEY**: Remove learned embeddings for prefix/suffix/ending → use deterministic features instead
4. Build training loop with checkpointing and early stopping
5. Create evaluation framework (accuracy, F1, reasoning quality)
6. Implement AST-to-AST training paradigm
7. Add attention visualization for interpretability
8. Hyperparameter tuning (model size, learning rate, batch size)

**Priority**: P0 (critical path)

**Blocking**: Requires Epic 1 completion (deterministic features)

**Related Milestones**: M2: Neural Reasoning Core

**Comment by marctjones** [comment]
## Updated Training Plan - AST-Native Architecture

Based on recent learnings, here's the clarified training approach for the Neural Reasoning Core:

### What We Train (On Pure Esperanto Corpus)

**Model 1: Root Embeddings (320K params)** ✅
- 5,000 Esperanto roots × 64d
- Training: Word2Vec on 26K corpus + Tatoeba similarity oracle
- Time: ~1 hour on CPU
- Output: Semantic similarity of roots only

**Model 2: AST Encoder (15-20M params)** ✅
- Graph Transformer on AST structure
- Training: Tatoeba EN-EO pairs (271K) as similarity oracle
- Time: 1-2 days on single GPU
- Purpose: Encode ASTs for retrieval/matching

**Model 3: AST Reasoning Core (20-100M params)** ⭐ THIS EPIC
- Graph-to-Graph Transformer
- Input: Query AST + Context ASTs → Output: Answer AST
- Training data: Generate Q&A, inference, paraphrase from 26K sentences
- Training objectives:
  - Question Answering (generate from corpus)
  - Inference/Entailment (synthetic from parsed sentences)
  - Slot Filling (mask AST nodes, predict from context)
  - Paraphrase (from Tatoeba translations)
- Time: 3-5 days on single GPU

### Key Insight: Why 26K Sentences is Enough

Traditional LLMs need 100M+ sentences because they learn grammar + semantics + reasoning.

Klareco needs only 26K because:
- **Grammar = 0 params** (deterministic from Esperanto rules)
- **Vocab = 5K roots** (vs 50K tokens, 10x smaller)
- **Learning = reasoning only** (on AST structure)

Math: 5K roots × 5-10 examples = 25K-50K sentences needed ✅

### Training Data Generation

From 26,725 sentences, generate:
1. **Q&A pairs**: `Kiu vidas la katon?` → `La hundo`
2. **Inference**: Premise + Hypothesis → Entailment/Contradiction
3. **Slot filling**: Mask AST nodes, predict from context
4. **Paraphrase**: Use Tatoeba translations

**Total training examples**: 130K-267K (5-10 variations per sentence)

### Architecture

```
Input: Query AST + Context ASTs (from retrieval)
   ↓
AST Encoder (Graph Transformer, 15M params)
   ↓
Cross-Attention over context
   ↓
Graph-to-Graph Decoder (20M params)
   ↓
Output: Answer AST
   ↓
Deparser (deterministic, 0 params)
   ↓
Esperanto answer (100% grammatically correct)
```

### Implementation Status

✅ Parser generates rich ASTs (91.8% parse rate)
✅ Corpus ready (26,725 sentences)
✅ AST-to-graph converter implemented
⬜ Root embedding training script
⬜ Q&A pair generation from corpus
⬜ AST reasoning core architecture
⬜ Training pipeline

### Documentation

See new wiki page: [AST-Native Training](https://github.com/marctjones/klareco/wiki/AST-Native-Training)

This clarifies **what to train**, **how to generate training data**, and **why our corpus is sufficient**.
